subsys/fs/nvs/nvs.c

```diff
@@ -10,161 +10,126 @@
+/* flash routines */
+/* basic aligned flash write to nvs address */
+int _nvs_flash_al_wrt(struct nvs_fs *fs, u32_t addr, const void *data,
+		      size_t len)
 {
-	if (len < 0x80) {
-		return get_wbs_aligned_len(fs, len + sizeof(u16_t)
-					   + 2*sizeof(u8_t));
-	}
-	return get_wbs_aligned_len(fs, len + sizeof(u16_t)
-				   + 2*sizeof(u16_t));
-}
+	int rc;
+	off_t offset;
+	size_t blen;
+	u8_t buf[fs->write_block_size];
 
-void _nvs_addr_advance(struct nvs_fs *fs, off_t *addr, size_t step)
-{
-	off_t fs_size = fs->sector_count * fs->sector_size;
+	offset = fs->offset;
+	offset += fs->sector_size * (addr >> 16);
+	offset += addr & 0xffff;
 
-	*addr += step;
-	if (*addr >= fs_size) {
-		*addr -= fs_size;
+	rc = flash_write_protection_set(fs->flash_device, 0);
+	if (rc) {
+		/* flash protection set error */
+		return rc;
 	}
-}
-void _nvs_addr_decrease(struct nvs_fs *fs, off_t *addr, size_t step)
-{
-	*addr -= step;
-	if (*addr < 0) {
-		*addr += (fs->sector_count * fs->sector_size);
+	blen = len & ~(fs->write_block_size);
+	rc = flash_write(fs->flash_device, offset, data, blen);
+	if (rc) {
+		/* flash write error */
+		return rc;
 	}
-}
-
-/* length is encoded in:
- *  1 byte if len < 128 byte: 0 b6 .. b0
- *  2 byte if len >= 128 byte 1 b14 .. b8 | b7 .. b0
- * _nvs_decode_length returns the number of byte used and updates the length
- * _nvs_decode_length is used to determine if the flash is unwritten, therefor
- * _nvs_decode_length returns 0xffff if the data in buf[0] and buf[1] is 0xff.
- */
-size_t _nvs_decode_length(u8_t *buf, size_t *len)
-{
-	if (buf[0] & 0x80) {
-		if (buf[0] == 0xff && buf[1] == 0xff) {
-			return 0xffff;
+	len -= blen;
+	offset += blen;
+	data += blen;
+	if (len) {
+		memcpy(buf, data, len);
+		memset(buf+len, 0xff, fs->write_block_size - len);
+		rc = flash_write(fs->flash_device, offset, buf,
+				 fs->write_block_size);
+		if (rc) {
+			/* flash write error */
+			return rc;
 		}
-		*len = (buf[1] & 0xff) | ((buf[0]&0x7f) << 8);
-		return sizeof(u16_t);
 	}
-	*len = buf[0];
-	return sizeof(u8_t);
+	(void) flash_write_protection_set(fs->flash_device, 1);
+	return 0;
 }
-/* length is encoded in:
- *  1 byte if len < 128 byte
- *  2 byte if len >= 128 byte
- * _nvs_encode_length returns the number of byte used and fills the buffer
- * _nvs_encode_length returns 0xffff if the len is larger then NVS_MAX_LEN
- */
-size_t _nvs_encode_length(u8_t *buf, size_t len)
+
+/* basic flash read from nvs address */
+int _nvs_flash_rd(struct nvs_fs *fs, u32_t addr, void *data,
+		  size_t len)
 {
-	if (len < 0x80) {
-		buf[0] = len;
-		return sizeof(u8_t);
-	}
-	if (len > NVS_MAX_LEN) {
-		return 0xffff;
-	}
-	buf[1] = len & 0xff;
-	buf[0] = ((len >> 8) & 0x7f) | 0x80;
-	return sizeof(u16_t);
-}
+	int rc;
+	off_t offset;
+
+	offset = fs->offset;
+	offset += fs->sector_size * (addr >> 16);
+	offset += addr & 0xffff;
 
-/* end basic calculation routines */
+	rc = flash_read(fs->flash_device, offset, data, len);
```

Comment:
Is it ensured that read operations are aligned to block-size? It is not expected form a flash driver that it support byte alignment while reading.